sympy/polys/polymatrix.py

```diff
@@ -1,87 +1,229 @@
-                ring = ring.unify(domain[i])
-        self.ring = ring
-
-    def _eval_matrix_mul(self, other):
-        self_cols = self.cols
-        other_rows, other_cols = other.rows, other.cols
-        other_len = other_rows*other_cols
-        new_mat_rows = self.rows
-        new_mat_cols = other.cols
-
-        new_mat = [0]*new_mat_rows*new_mat_cols
-
-        if self.cols != 0 and other.rows != 0:
-            mat = self._mat
-            other_mat = other._mat
-            for i in range(len(new_mat)):
-                row, col = i // new_mat_cols, i % new_mat_cols
-                row_indices = range(self_cols*row, self_cols*(row+1))
-                col_indices = range(col, other_len, other_cols)
-                vec = (mat[a]*other_mat[b] for a,b in zip(row_indices, col_indices))
-                # 'Add' shouldn't be used here
-                new_mat[i] = sum(vec)
-
-        return self.__class__(new_mat_rows, new_mat_cols, new_mat, copy=False)
-
-    def _eval_scalar_mul(self, other):
-        mat = [Poly(a.as_expr()*other, *a.gens) if isinstance(a, Poly) else a*other for a in self._mat]
-        return self.__class__(self.rows, self.cols, mat, copy=False)
-
-    def _eval_scalar_rmul(self, other):
-        mat = [Poly(other*a.as_expr(), *a.gens) if isinstance(a, Poly) else other*a for a in self._mat]
-        return self.__class__(self.rows, self.cols, mat, copy=False)
 
+    def __new__(cls, *args):
+        if isinstance(args[0], list):
+            elements, args = args[0], args[1:]
+            if isinstance(elements[0], (list, tuple)):
+                rows, cols = len(elements), len(elements[0])
+                items = [e for row in elements for e in row]
+            else:
+                rows, cols = len(elements), 1
+                items = elements
+        elif [type(a) for a in args[:3]] == [int, int, list]:
+            rows, cols, items, args = args[0], args[1], args[2], args[3:]
+        elif [type(a) for a in args[:3]] == [int, int, type(lambda: 0)]:
+            rows, cols, func, args = args[0], args[1], args[2], args[3:]
+            items = [func(i, j) for i in range(rows) for j in range(cols)]
+        else:
+            1/0
+        return cls.from_list(rows, cols, items, *args)
+
+    @classmethod
+    def from_list(cls, rows, cols, items, gens=None):
+        items = [sympify(item) for item in items]
+        if all(isinstance(item, Expr) for item in items):
+            polys = [Poly(item, gens) for item in items]
+        elif all(isinstance(item, Poly) for item in items):
+            polys = items
+        else:
+            1/0
+        if polys:
+            p = polys[0]
+            for p2 in polys[1:]:
+                p, _ = p.unify(p2)
+        else:
+            p = Poly(0, gens)
+        if gens is not None:
+            if not isinstance(gens, tuple):
+                gens = (gens,)
+            assert gens == p.gens
+        domain = p.domain.get_field()
```

Comment:
This is good for `gruntz` where `gens` are given, but if the items are Poly elements and there are no gens, then this is not what the user expects.

Reply:
I'm not sure I understand what you mean. A `Poly` always has gens.

Perhaps the domain should not always be converted to a field though.

For Risch it seems that all polys are over fields so you have something like `QQ(x)[t]`. PolyMatrix (on master) doesn't really distinguish between rational function fields of the ground domain and polynomials so operations like rref use division and implicitly convert from Poly to Expr. Here by ensuring that the domain is always a polynomial ring over a field we know that rref is always possible if all elements are ground which does seem to be the case whenever rref is used in Risch.

Maybe that only needs to be checked in the rref and nullspace functions though and we can leave the ground domain as it is everywhere else. Also the PolyMatrix tests use a `ring=` argument that is so far removed here and doesn't seem to be used anywhere outside of the PolyMatrix tests. Perhaps that should be added back.

Reply:
I was thinking of the default case `gens=None`. That would probably occur when PolyMatrix is used as a replacement for Matrix: No `gens` are given on the argument list.

Reply:
If this is called with Poly elements and no generators are given then the Poly are unified and those gens are taken as the gens of the PolyMatrix. There are several different cases:
```python
In [1]: from sympy.polys.polymatrix import PolyMatrix

In [2]: PolyMatrix([[Poly(x, x), Poly(y, y)]])  # gens from the Polys
Out[2]: PolyMatrix([[x, y]], (x, y))

In [3]: PolyMatrix([[x, y]], (x, y))           # gens given explicitly
Out[3]: PolyMatrix([[x, y]], (x, y))

In [4]: PolyMatrix([[x, y]])             # gens inferred from symbols
Out[4]: PolyMatrix([[x, y]], (x, y))

In [5]: PolyMatrix([[1, 2]])             # error just like Poly(1)
---------------------------------------------------------------------------
GeneratorsNeeded 
```
Which of those are you referring to?

Reply:
The first case. It looks like the `gens` parameter would be `None`. But maybe that does not matter provided there is at least one Poly element with gens.